File: modules/day-counter/android/src/main/java/expo/modules/daycounter/DayCounterModule.kt
package expo.modules.daycounter

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class DayCounterModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("DayCounter")

    Function("getLastResetDate") {
      DayCounterStore.getLastResetDate(appContext.reactContext ?: throw Exception("No context"))
    }

    Function("setLastResetDateToday") {
      val context = appContext.reactContext ?: throw Exception("No context")
      val date = DayCounterStore.setLastResetDateToday(context)
      DayCounterWidgetProvider.refreshAll(context)
      date
    }

    Function("getCurrentDayCount") {
      DayCounterStore.getCurrentDayCount(appContext.reactContext ?: throw Exception("No context"))
    }

    Function("refreshWidget") {
      DayCounterWidgetProvider.refreshAll(appContext.reactContext ?: throw Exception("No context"))
      true
    }
  }
}

File: src/native/dayCounterModule.ts
import { requireNativeModule } from 'expo-modules-core';

export type DayCounterNativeModule = {
  getLastResetDate(): string;
  setLastResetDateToday(): string;
  getCurrentDayCount(): number;
  refreshWidget(): boolean;
};

const DayCounter = requireNativeModule<DayCounterNativeModule>('DayCounter');

export async function getLastResetDate() {
  return DayCounter.getLastResetDate();
}

export async function setLastResetDateToday() {
  return DayCounter.setLastResetDateToday();
}

export async function getCurrentDayCount() {
  return DayCounter.getCurrentDayCount();
}

export async function refreshWidget() {
  return DayCounter.refreshWidget();
}
